great_expectations/core/expectation_suite.py

```diff
@@ -122,15 +128,34 @@ def __init__(  # noqa: PLR0913
 
         if expectations is None:
             expectations = []
-        expectation_configurations = [
-            ExpectationConfiguration(**expectation)
-            if isinstance(expectation, dict)
-            else expectation
-            for expectation in expectations
-        ]
-        self.expectations: list[Expectation] = [
-            self._build_expectation(config) for config in expectation_configurations
-        ]
+
+        if all(isinstance(exp, Expectation) for exp in expectations):
+            for expectation in expectations:
+                cast(Expectation, expectation)
+                if expectation.id:
+                    raise ValueError(
+                        "Expectations in parameter `expectations` must not belong to another ExpectationSuite. "
+                        "Instead, please use copies of Expectations, by calling `copy.copy(expectation)`."
+                    )
+                expectation.register_save_callback(save_callback=self._save_expectation)
+            self.expectations = expectations
+        else:
+            # legacy code path, remove once param `expectations` type is narrowed to just Expectation
+            for expectation in expectations:
+                if isinstance(expectation, Expectation):
```

Comment:
The need for this check is a side-effect of the type.

I'm guessing we want the runtime check as well, but if we don't allow mixing types, we should fix the type of `expectations`. It's more verbose, but I think we want

```
expectations: Optional[Union[
  Sequence[Expectation],
  Sequence[dict],
  Sequence[ExpectationConfiguration], # maybe these last two can get rolled in together since they are used interchangeably
]] = None,
```

Reply:
Alternatively, we could just handle mixed sequences. That actually might make some of the logic here simpler.